Introduce file_mounts_sync_continuously cluster option

alanwguo opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?
This PR introduces a feature to continuously sync file_mounts from the head_node to the worker_nodes. 
To enable this, this PR separates out the runtime_hash into two parts, a hash of the cluster configuration, and a hash of the contents of the file_mounts directories. If the configuration changes, the behavior remains the same, but if the contents of the file_mounts change and the continuous sync option is set, then the files will be synced to the worker nodes without re-running the init or setup commands.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28549/
Test FAILed.

AmeerHajAli commented on python/ray/autoscaler/gcp/example-full.yaml:
```diff
@@ -109,6 +109,9 @@ file_mounts: {
 #    "/path2/on/remote/machine": "/path2/on/local/machine",
 }
 
+# Whether changes to mounted files in the head node should sync to the worker node continuously
```
# Whether changes to mounted files in the head node should sync to the worker nodes continuously.

AmeerHajAli commented on python/ray/autoscaler/updater.py:
```diff
@@ -133,7 +137,10 @@ def do_update(self):
+        if node_tags.get(
+                TAG_RAY_RUNTIME_CONFIG) == self.runtime_hash and node_tags.get(
+                    TAG_RAY_FILE_MOUNTS_CONTENTS
+                ) == self.file_mounts_contents_hash:
```
Can we split this to two different cases (continuous vs continuous, it will also clean the if in the else)? or at least add a comment saying that we do not update self.file_mounts_contents_hash in noncontinuous file sync mode and thus it will always be equal to node_tags[TAG_RAY_FILE_MOUNTS_CONTENTS]. Can we not set these tags/check them if not in continuous sync mode?

alanwguo replied:
The contents of the file_mounts were hashed previously when continuous mode didn't even exist. I think its necessary so that that the autoscaler can detect changes to the file_mounts contents whenever the user runs `ray up`.

I prefer to add a comment vs separating out the two cases since the two cases are very similar except for 1 code path.

AmeerHajAli replied:
I see. Makes sense.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28653/
Test FAILed.

ericl commented on python/ray/autoscaler/autoscaler.py:
```diff
@@ -312,11 +323,18 @@ def launch_config_ok(self, node_id):
+        applied_config_hash = node_tags.get(TAG_RAY_RUNTIME_CONFIG)
+        applied_file_mounts_contents_hash = node_tags.get(
+            TAG_RAY_FILE_MOUNTS_CONTENTS)
+        if (applied_config_hash != self.runtime_hash
```
What's the reason we need to check the inner hash and not just the runtime one?

alanwguo replied:
So when the contents of the head node's file_mounts change, the files will be synced to the worker nodes.

This behavior should be the same as before since previously, the contents hash were included with the runtime hash.

ericl replied:
One behaviour I'm concerned with is whether "ray up" will still work if continuous_sync mode is on. If I understand correctly, that won't update the cluster if only files have changed right?

alanwguo replied:
ah I see, in that case the rsync will still happen but the setup commands wont run. I'll make a change so that it only doesn't run setup commands when update is called in the monitor.py flow.

alanwguo commented:
@ericl, can you help me merge this in? I'm not authorized.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28766/
Test FAILed.

ericl commented:
It looks like the lint job and test AutoScaler are failing in Travis.

On Wed, Jul 22, 2020, 6:20 PM UCB AMPLab <notifications@github.com> wrote:

> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28766/
> Test FAILed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/9544#issuecomment-662775822>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSVCTU4ZLV77CE5M3ADR46F7LANCNFSM4O6S3LBQ>
> .
>

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28804/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28809/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28817/
Test FAILed.

alanwguo commented:
> It looks like the lint job and test AutoScaler are failing in Travis.
> […](#)
> On Wed, Jul 22, 2020, 6:20 PM UCB AMPLab ***@***.***> wrote: Test FAILed. Refer to this link for build results (access rights to CI server needed): https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28766/ Test FAILed. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#9544 (comment)](https://github.com/ray-project/ray/pull/9544#issuecomment-662775822)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAADUSVCTU4ZLV77CE5M3ADR46F7LANCNFSM4O6S3LBQ> .

I fixed the lint issue. I also fixed a bug introduced in an unrelated PR to help fix the autoscaler test failures.

It seems, though, that there is a race condition with `autoscaler.update` competing with `node_launcher._launch_node` where a launched node would change the state of the node counts while `autoscaler.update` is being run. In ca7b27d, I attempt to reduce the chance of the race condition.

I wonder if we can just use a lock so `autoscaler.update` and `node_launcher._launch_node` can't run at the same time. It seems like it should be okay since `update` only runs every couple of seconds except in the unit tests.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28867/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28874/
Test FAILed.

alanwguo commented:
@ericl, can you help merge this for me now? Seems like the failing CI are unrelated to my change
